### PR TITLE
Tidy Cells

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR moves the storage of neighbours from the `Cell` class itself to `CellArray`, reducing memory requirements and data redundancy.

The main classes affected are `EnergyKernel` and `ForceKernel`, but since those classes already stored a reference to the correct `CellArray`, refactoring was not so bad. Removing the actual cell arrays from cells is a more involved job, and is left for a separate PR.

Other points to note:
- The `EnergyKernel` has been tidied in order to clarify the role of functions, make private some members, and prepare a little for the removal of `ProcessPool` from being stored in the object.
- There are still some speed increases in the `Molecule&`-based energy calculation, but not the 50% originally quoted. Turns out that it's better to calculate the energy *correctly* rather than quickly.
- For the `Atom&`-based energy calculation, there is a small reduction in performance using the new cell vector. However, the old routine was not threaded, so this penalty is offset somewhat by the now-parallel nature of the routine.

```
------------------------------------------------------------------------------------------------
MoleculeEnergy (EnergyKernel::energy(const Molecule &mol,...)) Time             CPU   Iterations
------------------------------------------------------------------------------------------------
<ProblemType::smallMolecule, Population::small>    OLD     0.193 ms        0.193 ms         3636
<ProblemType::smallMolecule, Population::small>    NEW     0.185 ms        0.185 ms         3788

<ProblemType::mediumMolecule, Population::small>   OLD      1.08 ms         1.08 ms          643
<ProblemType::mediumMolecule, Population::small>   NEW      1.01 ms         1.01 ms          696

<ProblemType::mediumMolecule, Population::medium>  OLD     0.893 ms        0.893 ms          749
<ProblemType::mediumMolecule, Population::medium>  NEW     0.859 ms        0.859 ms          827

-------------------------------------------------------------------------------------------------------------------------------
Atom / world energy (EnergyKernel::energy(const Atom &,...)    Time             CPU   Iterations
-------------------------------------------------------------------------------------------------------------------------------
<ProblemType::smallMolecule, Population::small>    OLD     55425 ns        55425 ns        12568
<ProblemType::smallMolecule, Population::small>    NEW     61518 ns        61518 ns        11394

<ProblemType::mediumMolecule, Population::small>   OLD     42291 ns        42289 ns        16143
<ProblemType::mediumMolecule, Population::small>   NEW     45792 ns        45791 ns        14953
```

### CODE REVIEW

File: web/content/docs/startingout/compilation.md [hunk 1]
```diff
@@ -174,19 +174,27 @@ Example: `-DBUILD_ANTLR_ZIPFILE:path=/path/to/antlr4-cpp-runtime-4.8-source.zip`
...
-Example: `-DBUILD_TESTS:bool=true`
+Usage: `-DBUILD_BECHMARKS:bool=true`
 
 Default: `false`
 
-#### `BUILD_BENCHMARKS`
+#### `BUILD_SYSTEM_TESTS`
 
-In addition to the main build, also build the benchmarks located in the benchmark directory. If this option is specified, the Google benchmark library will be downloaded and compiled by cmake. 
+In addition to the main build, also build system tests located in the `tests/` directory.
```
Comment (rprospero) [inline web/content/docs/startingout/compilation.md:187]:
```suggestion
In addition to the main build, also build system tests located in the `tests/` directory.  The system tests run dissolve on a large number of known data sets to confirm that there are no error in those calculations.
```

File: web/content/docs/startingout/compilation.md [hunk 2]
```diff
@@ -174,19 +174,27 @@ Example: `-DBUILD_ANTLR_ZIPFILE:path=/path/to/antlr4-cpp-runtime-4.8-source.zip`
...
 
-Usage: `-DBUILD_BECHMARKS:bool=true`
+Example: `-DBUILD_SYSTEM_TESTS:bool=true`
+
+Default: `false`
+
+#### `BUILD_UNIT_TESTS`
+
+In addition to the main build, also build unit tests located in the `unit/` directory.
```
Comment (rprospero) [inline web/content/docs/startingout/compilation.md:195]:
```suggestion
In addition to the main build, also build unit tests located in the `unit/` directory.  The unit tests check individual portions of the dissolve code base for problems.  These tests are not as comprehensive, but they run much quicker than the system tests and provide more precision as to where an bug is arising.
```

File: unit/test_cells.cpp [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+    water->addBond(1, 2);
+
+    // Setup Configuration
+    auto *cfg = dissolve.addConfiguration();
+    cfg->createBox({20, 20, 20}, {90, 90, 90});
+    cfg->cells().generate(cfg->box(), 7.0, dissolve.pairPotentialRange());
+    cfg->addMolecule(argon);
+    for (auto n = 0; n < 267; ++n)
+        cfg->addMolecule(water);
+    const std::vector<Vec3<double>> coords = {
```
Comment (rprospero) [inline unit/test_cells.cpp:62]: Not a big deal, but maybe `coords` could be declared as a a global and this definition could be moved to the end of the file?  It really breaks up the flow of the code.

**Review by rprospero — APPROVED** [review]
A couple of minor points, but otherwise nicely done.

File: src/classes/cellneighbour.h [hunk 4]
```diff
@@ -3,52 +3,30 @@
...
     {
-        // At the moment we find if a pair is already present through a linear seearch.
-        // Searching could be optimised using a set, but would cost more memory.
-        auto it = std::find(neighbours_.begin(), neighbours_.end(), neighbour);
-        if (it == neighbours_.end())
-        {
-            neighbours_.push_back(std::move(neighbour));
+        if (this->master_.index() == other.master_.index() && this->neighbour_.index() == other.neighbour_.index())
+            return true;
+        else if (this->neighbour_.index() == other.master_.index() && this->master_.index() == other.neighbour_.index())
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/classes/cellneighbour.h:27]: I did have a reason for doing it like this....  I think... Or maybe I didn't!  Comparing pointers is the logical thing to do.

File: unit/test_cells.cpp (hunk 3 above)
Comment (trisyoungs) [reply unit/test_cells.cpp]: Yep. Don't know why I didn't do that!

**Pull request merged**